tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSGen.scala

```diff
@@ -0,0 +1,298 @@
+              case BoxedFloatClass   => genCallHelper("isFloat", expr)
+              case BoxedDoubleClass  => typeof(expr) === "number"
+            }
+          } else {
+            className match {
+              case BoxedUnitClass    => genCallHelper("asUnit", expr)
+              case BoxedBooleanClass => genCallHelper("asBoolean", expr)
+              case BoxedByteClass    => genCallHelper("asByte", expr)
+              case BoxedShortClass   => genCallHelper("asShort", expr)
+              case BoxedIntegerClass => genCallHelper("asInt", expr)
+              case BoxedFloatClass   => genCallHelper("asFloat", expr)
+              case BoxedDoubleClass  => genCallHelper("asDouble", expr)
+            }
+          }
+        } else {
+          Apply(
+              envField(if (test) "is" else "as", className),
+              List(expr))
+        }
+
+      case ArrayType(base, depth) =>
+        Apply(
+            envField(if (test) "isArrayOf" else "asArrayOf", base),
+            List(expr, IntLiteral(depth)))
+    }
+  }
+
+  def genCallHelper(helperName: String, args: Tree*)(
+      implicit pos: Position): Tree = {
+    Apply(envField(helperName), args.toList)
+  }
+
+  def encodeClassVar(className: String)(
+      implicit pos: Position): Tree =
+    envField("c", className)
+
+  def genLoadModule(moduleClass: String)(
+      implicit pos: Position): Tree = {
+    import TreeDSL._
+    Apply(envField("m", moduleClass), Nil)
+  }
+
+  def genRawJSClassConstructor(className: String)(
+      implicit globalKnowledge: GlobalKnowledge, pos: Position): Tree = {
+
+    genRawJSClassConstructor(className,
+        globalKnowledge.getJSNativeLoadSpec(className))
+  }
+
+  def genRawJSClassConstructor(className: String,
+      spec: Option[irt.JSNativeLoadSpec])(
+      implicit pos: Position): Tree = {
+    spec match {
+      case None =>
+        // This is a Scala.js-defined JS class, call its class value accessor
+        Apply(envField("a", className), Nil)
+
+      case Some(spec) =>
+        genLoadJSFromSpec(spec)
+    }
+  }
+
+  def genLoadJSFromSpec(spec: irt.JSNativeLoadSpec)(
+      implicit pos: Position): Tree = {
+
+    def pathSelection(from: Tree, path: List[String]): Tree = {
+      path.foldLeft(from) {
+        (prev, part) => genBracketSelect(prev, StringLiteral(part))
+      }
+    }
+
+    spec match {
+      case irt.JSNativeLoadSpec.Global(path) =>
+        pathSelection(envField("g"), path)
+
+      case irt.JSNativeLoadSpec.Import(module, path) =>
+        val moduleValue = envModuleField(module)
+        path match {
+          case DefaultExportName :: rest =>
+            val defaultField = genCallHelper("moduleDefault", moduleValue)
+            pathSelection(defaultField, rest)
+          case _ =>
+            pathSelection(moduleValue, path)
+        }
+    }
+  }
+
+  private final val DefaultExportName = "default"
+
+  def envModuleField(module: String)(
+      implicit pos: Position): VarRef = {
```

Comment:
Looks like this method header would fit on one line.

In general, a bunch of methods in `JSGen` have shorter headers than they used to, and their formatting should be revisited.